stream/stream.go

```diff
@@ -4,36 +4,36 @@ package stream
 // In DDD terms, it is the "aggregate root".
 //
 // In order for your domain type to implement Stream:
-//   * Embed Events
-//   * Implement `StreamID` returning a unique identifier (usually the same as your stream's internal ID).
-//   * Implement `PopEvents` that returns the events on the Events.
-//   * Implement `FromEvents` to apply events to your stream.
+//   - Embed pointer to Events queue.
+//   - Implement `StreamID` returning a unique identifier (usually the same as your stream's internal ID).
+//   - Implement `Events` to expose the Events queue.
+//   - Implement `WithEvents` that returns a new instance with the provided Events queue set.
 //
 // Then an EventStore will be able to store and load it.
 //
 // Example:
 //
-//     type User struct {
-//         events stream.Events[*User]
-//         id string
-//     }
+//	type User struct {
+//	    events *stream.Events[User]
+//	    id string
+//	}
 //
-//     func (u *User) StreamID() stream.ID {
-//         return stream.ID(u.id)
-//     }
+//	func (u User) StreamID() stream.ID {
+//	    return stream.ID(u.id)
+//	}
 //
-//     func (u *User) PopEvents() []stream.VersionedEvent[*User] {
-//         return u.events.PopEvents()
-//     }
+//	func (u User) Events() *stream.Events[User] {
+//	    return u.events
+//	}
 //
-//     func (u *User) FromEvents(events stream.Events[*User]) error {
-//         p.events = events
-//         return stream.ApplyAll(p)
-//     }
+//	func (u User) WithEvents(events *stream.Events[User]) User {
```

Comment:
This won't work without the pointer receiver. I suggest keeping it consistent across the methods.

Reply:
Oh, I just now realized this returns the User. Isn't this syntax a bit unexpected?

Reply:
As the documentation says above, we expect the method will return a modified user/T.

Reply:
Wouldn't it be enough to modify itself in place? Since it returns itself anyway?

Reply:
Sorry, I do not get what you mean. Could you please provide some examples?

Reply:
```go
func (p Postcard) WithEvents(events *stream.Events[Postcard]) Postcard {
 	p.events = events
 	return p
 }
```

would become:

```go
func (p *Postcard) WithEvents(events *stream.Events[Postcard]) {
 	p.events = events
 }
```

Reply:
Or another option:

```go
func (p Postcard) NewWithEvents(events *stream.Events[Postcard]) *Postcard {
 	return &Postcard{
            events: events,
        }
 }
```

Reply:
I think the second one makes the most sense to me, as having a "method constructor" feels quite weird.

Reply:
But then the whole effort will go to waste, as we will require `Stream[*T]`. That is unfortunately the compromise I had to do in order to simplify the generic type definition.

Reply:
Ok, I got the second option. Let me try!

Reply:
Oh, I see now 🤔 so the side effect is that the three methods must not have a pointer receiver.

How about this then?

```go
func (p Postcard) NewWithEvents(events *stream.Events[Postcard]) Postcard {
 	return Postcard{
            events: events,
        }
 }
```

It's up to the user anyway to define this func, but maybe this is the closes to being "idiomatic"?

Reply:
perhaps `FromEvents` makes more sense in this case. Since it should return a new struct, not modify the existing one

Reply:
Actually I think this one would work as well, if we make it return `*T` in the interface?
```go
func (p Postcard) NewFromEvents(events *stream.Events[Postcard]) *Postcard {
 	return &Postcard{
            events: events,
        }
 }
```